### PR TITLE
fix(coinex): watchTickers - new watchTickers implementation

### LINKED ISSUE #18015: CoinEx watchTicker(): maximum recursion depth exceeded while calling a Python object
### Operating System

Ubuntu 20.04.5 LTS

### Programming Languages

python

### CCXT Version

3.1.6

### Description

The problem may occur a few minutes or a few hours after starting the program:

```
2023-05-25 11:37:39,543 ERROR maximum recursion depth exceeded while calling a Python object: Traceback (most recent call last):
  File "main.py", line 366, in watch_ticker
    await exchange.watch_ticker(symbol)
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/pro/coinex.py", line 375, in watch_ticker
    return await self.watch_tickers([symbol], params)
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/pro/coinex.py", line 405, in watch_tickers
    return await self.watch_tickers(symbols, params)
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/pro/coinex.py", line 405, in watch_tickers
    return await self.watch_tickers(symbols, params)
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/pro/coinex.py", line 405, in watch_tickers
    return await self.watch_tickers(symbols, params)
  [Previous line repeated 981 more times]
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/pro/coinex.py", line 388, in watch_tickers
    type, params = self.handle_market_type_and_params('watchTickers', None, params)
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/async_support/base/exchange.py", line 1925, in handle_market_type_and_params
    defaultType = self.safe_string_2(self.options, 'defaultType', 'type', 'spot')
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/base/exchange.py", line 706, in safe_string_2
    return Exchange.safe_either(Exchange.safe_string, dictionary, key1, key2, default_value)
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/base/exchange.py", line 818, in safe_either
    value = method(dictionary, key1)
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/base/exchange.py", line 645, in safe_string
    return str(dictionary[key]) if Exchange.key_exists(dictionary, key) else default_value
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/base/exchange.py", line 623, in key_exists
    if hasattr(dictionary, 'getitem') and not isinstance(dictionary, str):
RecursionError: maximum recursion depth exceeded while calling a Python object

2023-05-25 11:37:39,602 ERROR Task exception was never retrieved
future: <Task finished name='Task-5' coro=<exchange_loop() done, defined at main.py:305> exception=SystemExit(None)>
Traceback (most recent call last):
  File "main.py", line 366, in watch_ticker
    await exchange.watch_ticker(symbol)
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/pro/coinex.py", line 375, in watch_ticker
    return await self.watch_tickers([symbol], params)
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/pro/coinex.py", line 405, in watch_tickers
    return await self.watch_tickers(symbols, params)
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/pro/coinex.py", line 405, in watch_tickers
    return await self.watch_tickers(symbols, params)
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/pro/coinex.py", line 405, in watch_tickers
    return await self.watch_tickers(symbols, params)
  [Previous line repeated 981 more times]
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/pro/coinex.py", line 388, in watch_tickers
    type, params = self.handle_market_type_and_params('watchTickers', None, params)
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/async_support/base/exchange.py", line 1925, in handle_market_type_and_params
    defaultType = self.safe_string_2(self.options, 'defaultType', 'type', 'spot')
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/base/exchange.py", line 706, in safe_string_2
    return Exchange.safe_either(Exchange.safe_string, dictionary, key1, key2, default_value)
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/base/exchange.py", line 818, in safe_either
    value = method(dictionary, key1)
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/base/exchange.py", line 645, in safe_string
    return str(dictionary[key]) if Exchange.key_exists(dictionary, key) else default_value
  File "/home/user/.local/lib/python3.8/site-packages/ccxt/base/exchange.py", line 623, in key_exists
    if hasattr(dictionary, 'getitem') and not isinstance(dictionary, str):
RecursionError: maximum recursion depth exceeded while calling a Python object

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/asyncio/runners.py", line 47, in run
    _cancel_all_tasks(loop)
  File "/usr/lib/python3.8/asyncio/runners.py", line 62, in _cancel_all_tasks
    loop.run_until_complete(
  File "/usr/lib/python3.8/asyncio/base_events.py", line 603, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.8/asyncio/base_events.py", line 570, in run_forever
    self._run_once()
  File "/usr/lib/python3.8/asyncio/base_events.py", line 1859, in _run_once
    handle._run()
  File "/usr/lib/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "main.py", line 328, in exchange_loop
    await asyncio.gather(watch_balance(exchange), *tickers_loops, *loops)
  File "/usr/lib/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.8/asyncio/base_events.py", line 603, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.8/asyncio/base_events.py", line 570, in run_forever
    self._run_once()
  File "/usr/lib/python3.8/asyncio/base_events.py", line 1859, in _run_once
    handle._run()
  File "/usr/lib/python3.8/asyncio/events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "main.py", line 375, in watch_ticker
    exit()
  File "/usr/lib/python3.8/_sitebuiltins.py", line 26, in call
    raise SystemExit(code)
SystemExit: None
```

### Code

```
     exchange = getattr(ccxt.pro, 'coinex')()
     symbol = 'NEXA/USDT'
   
   while True:
        try:
            await exchange.watch_ticker(symbol)

        except NetworkError as e:
            logging.error(f'Network error: {exchange} watch ticker {symbol}: {str(e)}: %s', traceback.format_exc())

        except Exception as e:
            logging.error(f'{exchange} watch ticker {symbol}: {str(e)}: %s', traceback.format_exc())

            exit()

```

### PR DESCRIPTION
**Pull request by pcriadoperez**
There are two issues with the current watchTickers implementation

- it can hit amaximum recursion depth error (Fix #18015)
- if a user calls watchTickers (['BTC/USDT']) and watchTickers (['BTC/USDT', ETH/USDT']) and a new BTC/USDT ticker is received only the first ticker will resolve, when I believe on userland it would make sense for both to resolve. (It's been a while since I've tested this but I believe it is still an issue)

This PR proposes the following fix:
1. Add the symbols to the messageHash
2. When a message is received find all messageHashes that contain the symbol to resolve
3. Build array of tickers for that message hash to resolve.
4. Resolve each messageHash
5. In watchTickers check if to return only new updates or all tickers filtered by that symbol

Alternative solution. Use a new cache object to handle this like the one proposed for `watchPositions` in this PR #15622

If we decide this as a good solution it should then be applied to the rest of `watchTickers` implementations.

### CODE REVIEW

File: ts/src/base/Exchange.ts [hunk 1]
```diff
@@ -1358,6 +1358,18 @@ export default class Exchange {
     // ------------------------------------------------------------------------
     // METHODS BELOW THIS LINE ARE TRANSPILED FROM JAVASCRIPT TO PYTHON AND PHP
 
+    findMessageHashes (client, element: string): string[] {
```
Comment (carlosmiei) [inline ts/src/base/Exchange.ts:1361]: @pcriadoperez there is a lengthy discussion regarding the best way to handle `watchTickers`. I also believe this is the easiest way of making it right without needing an dispatcher/subscription-manager. The only downside I see is the necessity of doing a lot of string comparisons (with the `indexOf`), but in practice and unless the user specifies hundreds of symbols it should be ok I guess.

Comment (carlosmiei) [reply ts/src/base/Exchange.ts:1361]: (actually I did something similar in `whitebit` but was eventually removed)

File: ts/src/pro/coinex.ts [hunk 2]
```diff
@@ -141,10 +141,21 @@ export default class coinex extends coinexRest {
...
             const parsedTicker = this.parseWSTicker (rawTicker, market);
-            const messageHash = 'ticker:' + symbol;
             this.tickers[symbol] = parsedTicker;
             newTickers.push (parsedTicker);
-            client.resolve (parsedTicker, messageHash);
+        }
+        const messageHashes = this.findMessageHashes (client, 'tickers:');
+        for (let i = 0; i < messageHashes.length; i++) {
+            const messageHash = messageHashes[i];
+            const parts = messageHash.split (':');
```
Comment (carlosmiei) [inline ts/src/pro/coinex.ts:150]: @pcriadoperez hum we can't use `:` here safely, what if you have `messageHas = 'tickers:BTC/USDT,BTC/USDT:USDT'` 

Then parts would be `['tickers', 'BTC/USDT,BTC/USDT', 'USDT']`

File: ts/src/base/Exchange.ts (hunk 1 above)
Comment (pcriadoperez) [reply ts/src/base/Exchange.ts:1361]: Yes, I think that could be the only issue, but I believe it should be fast as it's stored in memory.
If we wanted to get around it maybe a more efficient way would be to store a hashmap that indexes the futures by symbol. But then we would be introducing another object and more complexity, and not sure if it's worth it for this case.

File: ts/src/pro/coinex.ts (hunk 2 above)
Comment (pcriadoperez) [reply ts/src/pro/coinex.ts:150]: good catch.. I replaced it for '::'

File: ts/src/base/Exchange.ts (hunk 1 above)
Comment (carlosmiei) [reply ts/src/base/Exchange.ts:1361]: @pcriadoperez Yeah let's keep it as is for now

**Comment by carlosmiei** [comment]
LGTM ✅ 

```
 p coinex watchTickers '["BTC/USDT", "LTC/USDT"]'   
Python v3.10.9
CCXT v3.1.8
coinex.watchTickers(['BTC/USDT', 'LTC/USDT'])
{'BTC/USDT': {'ask': None,
              'askVolume': 14.42065195,
              'average': 26958.91,
              'baseVolume': 712.73668029,
              'bid': None,
              'bidVolume': 15.8248204,
              'change': 602.26,
              'close': 27260.04,
              'datetime': None,
              'high': 27350.0,
              'info': {'buy_total': '15.82482040',
                       'close': '27260.04',
                       'deal': '19273673.1639044680',
                       'high': '27350.00',
                       'last': '27260.04',
                       'low': '26617.27',
                       'open': '26657.78',
                       'period': 86400,
                       'sell_total': '14.42065195',
                       'volume': '712.73668029'},
              'last': 27260.04,
              'low': 26617.27,
              'open': 26657.78,
              'percentage': 2.259227887693574,
              'previousClose': None,
              'quoteVolume': 19273673.16390447,
              'symbol': 'BTC/USDT',
              'timestamp': None,
              'vwap': 27041.78653477235},
 'LTC/USDT': {'ask': None,
              'askVolume': 306.35903133,
              'average': 88.485,
              'baseVolume': 1978.4694196,
              'bid': None,
              'bidVolume': 336.9417756,
              'change': 1.41,
              'close': 89.19,
              'datetime': None,
              'high': 90.37,
              'info': {'buy_total': '336.94177560',
                       'close': '89.19',
                       'deal': '176318.7235619935',
                       'high': '90.37',
                       'last': '89.19',
                       'low': '87.39',
                       'open': '87.78',
                       'period': 86400,
                       'sell_total': '306.35903133',
                       'volume': '1978.46941960'},
              'last': 89.19,
              'low': 87.39,
              'open': 87.78,
              'percentage': 1.6062884483937114,
              'previousClose': None,
              'quoteVolume': 176318.7235619935,
              'symbol': 'LTC/USDT',
              'timestamp': None,
              'vwap': 89.11875099774906}}
```

**Review by carlosmiei — APPROVED** [review]

**Pull request merged**